File: scripts/ultrawork.mjs
#!/usr/bin/env node
/**
 * Ultrawork — 一键三连。顺序执行：Go 编译 → 前端构建 → 测试 → 红线审查 → Git 状态。
 * 由 scripts/ultrawork.py 迁移（2026-08-03），逻辑逐点保真。
 */
import { execFileSync } from 'node:child_process';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '..');
const PASS = '[OK]';
const FAIL = '[FAIL]';

function run(cmd, cwd = ROOT, label = '', stopOnFail = true, tail = 10) {
  process.stdout.write(`\n=== ${label} ===\n`);
  try {
    const stdout = execFileSync(cmd[0], cmd.slice(1), { cwd, encoding: 'utf-8', timeout: 180000 });
    process.stdout.write(`  ${PASS} ${label} passed\n`);
    return true;
  } catch (e) {
    const out = (e.stdout || '') + (e.stderr || '');
    process.stdout.write(`  ${FAIL} ${label} failed (showing last ${tail} lines)\n`);
    for (const line of out.trim().split('\n').slice(-tail)) {
      process.stdout.write(`    ${line}\n`);
    }
    if (stopOnFail) process.exit(1);
    return false;
  }
}

console.log('========== Ultrawork ==========');

run(['go', 'build', './go/...'], ROOT, '[1/5] Go Build');
run(['npx', 'vite', 'build'], path.join(ROOT, 'frontend'), '[2/5] Frontend Build');
run(['go', 'test', './go/...', '-count=1'], ROOT, '[3/5] Go Test');
run(['node', 'scripts/review.mjs'], ROOT, '[4/5] Code Review', false);
run(['git', 'status', '--short'], ROOT, '[5/5] Git Status', false);

console.log(`\n${PASS} Ultrawork complete`);

File: scripts/ultrawork.py
#!/usr/bin/env python3
"""Ultrawork — 一键三连。顺序执行：Go 编译 → 前端构建 → 测试 → 红线审查 → Git 状态。"""
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
PASS = "[OK]"
FAIL = "[FAIL]"


def run(cmd, cwd=None, label="", stop_on_fail=True, tail=10):
    cwd = cwd or ROOT
    sys.stdout.buffer.write(f"\n=== {label} ===\n".encode("utf-8"))
    try:
        r = subprocess.run(cmd, cwd=cwd, capture_output=True, timeout=180)
        out = r.stdout.decode("utf-8", errors="replace") + r.stderr.decode("utf-8", errors="replace")
        if r.returncode == 0:
            sys.stdout.buffer.write(f"  {PASS} {label} passed\n".encode("utf-8"))
            return True
        else:
            sys.stdout.buffer.write(f"  {FAIL} {label} failed (showing last {tail} lines)\n".encode("utf-8"))
            for line in out.strip().split("\n")[-tail:]:
                sys.stdout.buffer.write(f"    {line}\n".encode("utf-8"))
            if stop_on_fail:
                sys.exit(1)
            return False
    except FileNotFoundError:
        print(f"  {FAIL} command not found: {cmd[0]}")
        if stop_on_fail:
            sys.exit(1)
        return False
    except subprocess.TimeoutExpired:
        print(f"  {FAIL} timeout")
        if stop_on_fail:
            sys.exit(1)
        return False


def main():
    print("========== Ultrawork ==========")

    run(["go", "build", "./go/..."], label="[1/5] Go Build")
    run(["npx", "vite", "build"], ROOT / "frontend", label="[2/5] Frontend Build")
    run(["go", "test", "./go/...", "-count=1"], label="[3/5] Go Test")
    run(["python3", "scripts/review.py"], stop_on_fail=False, label="[4/5] Code Review")
    run(["git", "status", "--short"], stop_on_fail=False, label="[5/5] Git Status")

    print(f"\n{PASS} Ultrawork complete")


if __name__ == "__main__":
    main()
